File: BMWCenter/Core/Care/FuelTrimMonitor.swift
import Foundation
import SwiftData

@MainActor
final class FuelTrimMonitor: CareFeature {
    let id = "fuelTrim"
    let requiredPIDs: Set<UInt8> = [0x06, 0x07]
    let optionalPIDs: Set<UInt8> = [0x08, 0x09]

    private let baseline: BaselineLearner
    private let modelContext: ModelContext

    private struct BandHit: Sendable {
        var tripIDs: Set<UUID> = []
        var durationS: Double = 0
        var leanOrRich: String?
    }

    private var sessionHits: [String: BandHit] = [:]
    private var bandAccum: [String: (sum: Double, n: Int)] = [:]

    init(baseline: BaselineLearner, modelContext: ModelContext) {
        self.baseline = baseline
        self.modelContext = modelContext
    }

    func isEnabled(settings: AppSettings) -> Bool { settings.careFuelTrimMonitor }

    func evaluate(snapshot: VehicleSnapshot, context: inout CareContext) -> [CareCue] {
        guard let ltft = snapshot.ltftBank1 else { return [] }
        let stft = snapshot.stftBank1 ?? 0
        let total = ltft + stft
        let coolant = snapshot.coolantC ?? 0
        let fuel = snapshot.fuelLevelPct ?? 100
        let load = snapshot.engineLoadPct ?? 0
        let speed = snapshot.speedKmh ?? 0

        if let refuel = context.lastRefuelAt,
           context.now.timeIntervalSince(refuel) < 600 {
            return []
        }
        guard coolant > 80, fuel > 15 else { return [] }

        let band: String
        switch (speed, load) {
        case (...3, ...25): band = "idle"
        case (_, 25...65): band = "part"
        case (_, 65...): band = "high"
        default: return []
        }

        baseline.observe(
            key: "ltft.\(band)",
            value: total,
            bucketKey: "",
            minSamples: 60,
            range: -40...40,
            now: context.now
        )
        var acc = bandAccum[band] ?? (0, 0)
        acc.sum += total
        acc.n += 1
        bandAccum[band] = acc

        return []
    }

    func onTripEnded(trip: Trip, context: CareContext) -> [CareCue] {
        defer { bandAccum.removeAll() }
        guard trip.durationS >= 60 else { return [] }

        let idleAvg = avg("idle")
        let partAvg = avg("part")
        let highAvg = avg("high")

        var pattern: String?
        if let idle = idleAvg, idle > 12, (partAvg ?? 0) < 12, (highAvg ?? 0) < 12 {
            pattern = "lean_idle"
        } else if let idle = idleAvg, idle > 12, (partAvg ?? 0) > 12 {
            pattern = "lean_all"
        } else if let high = highAvg, high > 15, (idleAvg ?? 0) < 12 {
            pattern = "lean_high"
        } else if let idle = idleAvg, idle < -12, (partAvg ?? 0) < -12 {
            pattern = "rich_all"
        }

        // Bank imbalance
        // (snapshot not available at trip end — skip live bank check here)

        guard let pattern else { return [] }

        var hit = sessionHits[pattern] ?? BandHit()
        hit.tripIDs.insert(trip.id)
        hit.durationS += trip.durationS
        hit.leanOrRich = pattern.contains("rich") ? "rich" : "lean"
        sessionHits[pattern] = hit

        guard hit.tripIDs.count >= 3, hit.durationS >= 20 * 60 else { return [] }

        modelContext.insert(ProtectionEvent(
            typeRaw: "fuelTrim",
            severityRaw: "alarm",
            value: idleAvg ?? partAvg ?? highAvg ?? 0,
            thresholdUsed: 12
        ))
        try? modelContext.save()

        let drifting = LocalizedCare.string("trim.drifting")
        return [CareCue(
            id: "trim.drift",
            text: drifting,
            severity: .coach,
            localizationKey: "trim.drifting"
        )]
    }

    nonisolated static func validateAlert(tripCount: Int, durationS: Double, postRefuelS: Double) -> Bool {
        tripCount >= 3 && durationS >= 20 * 60 && postRefuelS >= 600
    }

    private func avg(_ band: String) -> Double? {
        guard let acc = bandAccum[band], acc.n > 0 else { return nil }
        return acc.sum / Double(acc.n)
    }
}

File: mobile/src/core/care/fuelTrimMonitor.ts
import { db } from "../storage/db";
import { protectionEvents } from "../storage/schema";
import { baselineLearner } from "./baselineLearner";
import type { VehicleSnapshot } from "../obd/vehicleSnapshot";
import type { Trip } from "../storage/models";
import type { CareCue, CareContext } from "./careTypes";
import type { CareFeature } from "./careFeature";

interface BandHit {
  tripIds: Set<string>;
  durationS: number;
}

const THERMOSTAT_OPEN_C = 88;

/** Watches long-term fuel trim for a
 * lean/rich drift pattern across trips (skipping the unported ethanol/altitude
 * bucket adjustments from the vehicle-profile system). */
class FuelTrimMonitor implements CareFeature {
  id = "fuelTrim";

  private bandAccum: Record<string, { sum: number; n: number }> = {};
  private sessionHits: Record<string, BandHit> = {};

  isEnabled(settings: { careFuelTrimMonitor: boolean }): boolean {
    return settings.careFuelTrimMonitor;
  }

  evaluate(snapshot: VehicleSnapshot, context: CareContext): CareCue[] {
    if (snapshot.ltftBank1 == null) return [];
    const stft = snapshot.stftBank1 ?? 0;
    const total = snapshot.ltftBank1 + stft;
    const coolant = snapshot.coolantC ?? 0;
    const fuel = snapshot.fuelLevelPct ?? 100;
    const normLoad = snapshot.engineLoadPct ?? 0;
    const speed = snapshot.speedKmh ?? 0;
    const rpm = snapshot.rpm ?? 0;

    if (coolant <= THERMOSTAT_OPEN_C - 5 || fuel <= 15) return [];
    if (normLoad < 5 && rpm > 1200) return []; // decel fuel-cutoff proxy — trim not representative

    let band: string;
    if (speed <= 3 && normLoad <= 25) band = "idle";
    else if (normLoad >= 25 && normLoad <= 65) band = "part";
    else if (normLoad >= 65) band = "high";
    else return [];

    baselineLearner.observe(`ltft.${band}`, total, "", 60, [-40, 40], context.now).catch(() => undefined);
    const acc = this.bandAccum[band] ?? { sum: 0, n: 0 };
    acc.sum += total;
    acc.n += 1;
    this.bandAccum[band] = acc;

    return [];
  }

  onTripEnded(trip: Trip, context: CareContext): CareCue[] {
    // Taken before the reset, not read through `this` afterwards: the closure
    // used to resolve `this.bandAccum` at call time, by which point it had
    // already been emptied two lines above, so every average came back
    // undefined and no drift pattern could ever match.
    const accum = this.bandAccum;
    this.bandAccum = {};
    const avg = (band: string) => {
      const acc = accum[band];
      return acc && acc.n > 0 ? acc.sum / acc.n : undefined;
    };
    if (trip.durationS < 60) return [];

    const idleAvg = avg("idle");
    const partAvg = avg("part");
    const highAvg = avg("high");

    let pattern: string | undefined;
    if (idleAvg != null && idleAvg > 20 && (partAvg ?? 0) <= 10 && (highAvg ?? 0) <= 10) pattern = "lean_idle";
    else if (idleAvg != null && idleAvg > 20 && (partAvg ?? 0) > 20) pattern = "lean_all";
    else if (highAvg != null && highAvg > 20 && (idleAvg ?? 0) <= 10) pattern = "lean_high";
    else if (idleAvg != null && idleAvg < -20 && (partAvg ?? 0) < -20) pattern = "rich_all";
    if (!pattern) return [];

    const hit = this.sessionHits[pattern] ?? { tripIds: new Set<string>(), durationS: 0 };
    hit.tripIds.add(trip.id);
    hit.durationS += trip.durationS;
    this.sessionHits[pattern] = hit;

    if (hit.tripIds.size < 3 || hit.durationS < 20 * 60) return [];

    db.insert(protectionEvents)
      .values({
        id: `pe_${Date.now()}_${Math.random().toString(36).slice(2, 6)}`,
        type: "fuelTrim",
        severity: "alarm",
        t: Date.now(),
        value: idleAvg ?? partAvg ?? highAvg ?? 0,
        thresholdUsed: 20,
      })
      .catch(() => undefined);

    // Cleared once announced, so the pattern has to establish itself again
    // before it speaks. Without this the counters stay over the threshold and
    // every later trip with the same drift repeats the same cue forever.
    delete this.sessionHits[pattern];

    return [{ id: "trim.drift", text: "Fuel trim has been drifting outside normal range across recent trips.", severity: "coach" }];
  }

  resetTrip(): void {
    this.bandAccum = {};
  }
}

export const fuelTrimMonitor = new FuelTrimMonitor();
